src/java/picard/sam/revertsam/RevertSamUtil.java

```diff
@@ -0,0 +1,188 @@
+package picard.sam.revertsam;
+
+import htsjdk.samtools.BAMRecordCodec;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMFileWriter;
+import htsjdk.samtools.SAMFileWriterFactory;
+import htsjdk.samtools.SAMReadGroupRecord;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SAMRecordQueryNameComparator;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.ValidationStringency;
+import htsjdk.samtools.filter.FilteringIterator;
+import htsjdk.samtools.filter.SamRecordFilter;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.FastqQualityFormat;
+import htsjdk.samtools.util.Log;
+import htsjdk.samtools.util.QualityEncodingDetector;
+import htsjdk.samtools.util.SortingCollection;
+import picard.PicardException;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Functions for creating writers, headers, and other objects needed by RevertSam.
+ */
+public class RevertSamUtil {
+
+    private final static Log log = Log.getInstance(RevertSamUtil.class);
+
+    public static List<ReadGroup> makeReadGroupList(
+            final SAMFileHeader inHeader,
+            final OutputMap outputMap,
+            final SAMFileHeader.SortOrder sortOrder,
+            final boolean removeAlignmentInformation) {
+
+        final List<ReadGroup> readGroupList = new ArrayList<ReadGroup>();
+        for (final SAMReadGroupRecord samReadGroup : inHeader.getReadGroups()) {
+            final SAMFileHeader samHeader = createOutHeaderForReadGroup(samReadGroup, inHeader, sortOrder, removeAlignmentInformation);
+            final String id = samReadGroup.getId();
+            final ReadGroup readGroup = new ReadGroup(id, samHeader, outputMap.get(id));
+            readGroupList.add(readGroup);
+        }
+        return readGroupList;
+    }
+
+    private static SAMFileHeader createOutHeaderForReadGroup(
+            final SAMReadGroupRecord samReadGroup,
+            final SAMFileHeader inHeader,
+            final SAMFileHeader.SortOrder sortOrder,
+            final boolean removeAlignmentInformation) {
+
+        final SAMFileHeader outHeader = new SAMFileHeader();
+        outHeader.addReadGroup(samReadGroup);
+        outHeader.setSortOrder(sortOrder);
+        if (!removeAlignmentInformation) {
+            outHeader.setSequenceDictionary(inHeader.getSequenceDictionary());
+            outHeader.setProgramRecords(inHeader.getProgramRecords());
+        }
+        return outHeader;
+    }
+
+    public static RevertSamWriter makeSingleOrPerReadGroupWriter(
```

Comment:
This seems like a lot of extra code and effort to hide if we are writing to one output file or one per read group.  What does this buy you?  It's not like `RevertSam` would be too much more complex if you had `Map<String, SAMFileWriter> writers`.